src/test/suite/helper.ts

```diff
@@ -47,43 +80,813 @@ export function mockTextDocument(uri: Uri, languageId: string, source: string[])
+        });
+    }
+
+    public register(_data: RegistrationData<ExecuteCommandRegistrationOptions>): void {
+        // Do nothing. Otherwise we end up with double registration
+        traceInfo('Registering dummy command feature');
+    }
+
+    public unregister(id: string): void {
+        let disposables = this._commands.get(id);
+        if (disposables) {
+            disposables.forEach((disposable) => disposable.dispose());
+        }
+    }
+
+    public dispose(): void {
+        this._commands.forEach((value) => {
+            value.forEach((disposable) => disposable.dispose());
+        });
+        this._commands.clear();
+    }
+}
+
+async function startLanguageServer(languageServerFolder: string, pythonPath: string) {
+    const bundlePath = path.join(languageServerFolder, 'server.bundle.js');
+    const nonBundlePath = path.join(languageServerFolder, 'server.js');
+    const modulePath = (await fs.pathExists(nonBundlePath)) ? nonBundlePath : bundlePath;
+    const debugOptions = { execArgv: ['--nolazy', '--inspect=6600'] };
+    cancellationStrategy = new FileBasedCancellationStrategy();
+
+    // If the extension is launched in debug mode, then the debug server options are used.
+    const serverOptions: vslc.ServerOptions = {
+        run: {
+            module: bundlePath,
+            transport: vslc.TransportKind.ipc,
+            args: cancellationStrategy.getCommandLineArguments()
+        },
+        // In debug mode, use the non-bundled code if it's present. The production
+        // build includes only the bundled package, so we don't want to crash if
+        // someone starts the production extension in debug mode.
+        debug: {
+            module: modulePath,
+            transport: vslc.TransportKind.ipc,
+            options: debugOptions,
+            args: cancellationStrategy.getCommandLineArguments()
+        }
+    };
+
+    // Client options need to include our middleware piece
+    const clientOptions: vslc.LanguageClientOptions = {
+        documentSelector: PYTHON,
+        workspaceFolder: undefined,
+        synchronize: {
+            configurationSection: PYTHON_LANGUAGE
+        },
+        outputChannel: vscode.window.createOutputChannel('pylance-test'),
+        revealOutputChannelOn: RevealOutputChannelOn.Never,
+        initializationOptions: {
+            experimentationSupport: true
+        },
+        middleware: createMiddlewareAddon(
+            notebookApi,
+            () => languageClient,
+            traceInfo,
+            'python',
+            /.*\.(ipynb|interactive)/m,
+            pythonPath
+        ),
+        connectionOptions: {
+            cancellationStrategy
+        }
+    };
+
+    languageClient = new vslc.LanguageClient('lsp-middleware-test', serverOptions, clientOptions);
+
+    // Before starting do a little hack to prevent the pylance double command registration (working with Jake to have an option to skip commands)

```

Comment:
There's a problem with running two pylances at the same time though (one from the python extension and one from this test). Pylance registers commands. Commands can only be registered once. This hack here removes the command registration from the vscodelanguageclient module. 

@jakebailey and I are discussing having a setting in pylance to have pylance skip commands when run from here/jupyter.